[PIP 86] add doc for initialize function interface (#14315)
apache/pulsar: site2/docs/functions-develop.md

:---------|:------------|:---------
Language-native interface | No Pulsar-specific libraries or special dependencies required (only core libraries from Java/Python). | Functions that do not require access to the function [context](#context).
Pulsar Function SDK for Java/Python/Go | Pulsar-specific libraries that provide a range of functionality not provided by "native" interfaces. | Functions that require access to the function [context](#context).

> Extended Pulsar Function SDK for Java | An extension to Pulsar-specific libraries, providing the initialization and close interfaces in java. | Functions that require initializing and releasing external resources.

Reviewer: ```suggestion
Extended Pulsar Function SDK for Java | An extension to Pulsar-specific libraries, providing the initialization and close interfaces in Java. | Functions that require initializing and releasing external resources.
```

```
For complete code, see [here](https://github.com/apache/pulsar/blob/77cf09eafa4f1626a53a1fe2e65dd25f377c1127/pulsar-function-go/examples/inputFunc/inputFunc.go#L20-L36).

> 

Reviewer: ```suggestion
<!--END_DOCUSAURUS_CODE_TABS-->

### Extended Pulsar Function SDK for Java
```

> Function initialize and close interface are available in **Pulsar Function SDK for Java 2.10.0 or later**.

Reviewer: This extended Function SDK provides an interface to initialize and release external resources, which is easy and efficient to use and expand. Before starting a Function, you don't have to make various judgments in the Function interface for those resources which only need one-time initialization. After closing a Function, you don't need to manually close the referenced external resources. The order of releasing external resources is controlled by the Function itself.

> **Note** 
> The extended Function SDK for java is available in Pulsar 2.10.0 or later versions. Before using it, you need to set up Pulsar Functions worker 2.10.0 or later versions.

> Function initialize and close interface are available in **Pulsar Function SDK for Java 2.10.0 or later**.

Reviewer: ```suggestion
> **Note** 
> The extended Function SDK for java is available in Pulsar 2.10.0 or later versions. Before using it, you need to set up Pulsar Functions worker 2.10.0 or later versions.
```

> You can use these interface to process any resource initialize and release logic.

Reviewer: ```suggestion
This extended Function SDK provides an interface to initialize and release external resources, which is easy and efficient to use and expand. Before starting a Function, you don't have to make various judgments in the Function interface for those resources which only need one-time initialization. After closing a Function, you don't need to manually close the referenced external resources. The order of releasing external resources is controlled by the Function itself.
```

> The following example uses initialize and close interface.

Reviewer: ```suggestion
The following example uses the extended Function SDK to initialize and release the external resource RedisClient.
```

```Java
import org.apache.pulsar.functions.api.Context;
import org.apache.pulsar.functions.api.Function;

public class InitializableFunction implements Function<String, String> {
    private RedisClient client;

    @Override
    public void initialize(Context context) {
        Map<String, Object> connectInfo = context.getUserConfigMap();
        client=init(connectInfo);
    }
    
    @Override
    public String process(String input, Context context) {
        String value = client.get(key);
        return String.format("%s-%s", input, value);
    }

    @Override
    public void close() {
        client.close();
    }
}
```

> Notice that to use initialize interface, you should set up Pulsar Functions worker 2.10.0 or later.

Reviewer: This notice can be deleted because the information has been clarified in the beginning.

> <!--END_DOCUSAURUS_CODE_TABS-->

Reviewer: It should be the end of the last section.

}
```
For complete code, see [here](https://github.com/apache/pulsar/blob/77cf09eafa4f1626a53a1fe2e65dd25f377c1127/pulsar-function-go/examples/inputFunc/inputFunc.go#L20-L36).
<!--END_DOCUSAURUS_CODE_TABS-->

### Extended Pulsar Function SDK for Java

> This extended Pulsar Function SDK provides two addition interface to initialize and release external resources.

Reviewer: ```suggestion
This extended Pulsar Function SDK provides two additional interfaces to initialize and release external resources.
```

> By using initialize interface, you could do things like initialize external resources which only need one-time initialization when function instance starting.

Reviewer: ```suggestion
- By using the `initialize` interface, you can initialize external resources which only need one-time initialization when the function instance starts.
```

> By using close interface, you could close the referenced external resources when function instance closing. 

Reviewer: ```suggestion
- By using the `close` interface, you can close the referenced external resources when the function instance closes. 
```

> **Note**
>
> The extended Pulsar Function SDK for Java is available in Pulsar 2.10.0 and later versions.
> Before using it, you need to set up Pulsar Function worker 2.10.0 or later versions.

> The following example uses the extended interface of Pulsar Function SDK for Java to initialize RedisClient when function instance starting and release it when function instance closing.

Reviewer: ```suggestion
The following example uses the extended interface of Pulsar Function SDK for Java to initialize RedisClient when the function instance starts and release it when the function instance closes.
```
